File: app/src/main/java/com/PoeticManifestations/vicarioustexts/story/Story.java
package com.PoeticManifestations.vicarioustexts.story;

import java.util.Date;

/*
* Story class is used to contain a Story
* It has methods to Start Reading a Story
*/

public class Story {
    StoryMessage storyStart; //Starting node of the story
    StoryMessage currentMessage; //Points to a specific Message in the story
    private int totalMessages, progress;

    //Story info
    private String genre, author;
    Date publicationDate;

    public Story(String genre, String author) {
        publicationDate = null;
        this.genre = genre;
        this.author = author;
        this.storyStart = null;
        this.totalMessages = 0;
    }

    // Start / Reset Story progress
    public void startStory(){
        if (storyStart!=null) {
            currentMessage = storyStart;
            progress = 1;
        }
    }

    //Gets the StoryMessage currently being point to
    public StoryMessage getCurrentMessage(){
        return currentMessage;
    }

    //Will store next message in currentMessage
    //Returns true if a next message exists
    //false if next message is null
    public boolean nextMessage(){
        currentMessage = currentMessage.getNextMessage();
        progress++;
        return currentMessage != null;
    }

    //Will check if Story has a next message in the list
    public boolean hasNextMessage(){
        return currentMessage.getNextMessage() != null;
    }

    //Used to set the totalMessages in inherited class
    void setTotalMessages(int totalMessages){
        this.totalMessages = totalMessages;
    }

    //--------- Getters -------------

    //Get the total number of messages in the Story
    public int getTotalMessages(){
        return totalMessages;
    }

    public Date getPublicationDate(){
        //Can produce null pointer exception
        return this.publicationDate;
    }

    public String getAuthor(){
        return this.author;
    }

    public String getGenre(){
        return this.genre;
    }

}

File: app/src/main/java/com/poeticManifestations/vicarioustexts/story/Story.kt
package com.poeticManifestations.vicarioustexts.story

import java.util.*

/*
* Story class is used to contain a Story
* It has methods to Start Reading a Story
*/

open class Story(val genre: String = "Unknown", val author: String = "Anonymous") {

    protected lateinit var storyStart: StoryMessage   //Starting node of the story

    var currentMessage: StoryMessage? = null //Points to a specific Message in the story
        protected set

    //Total Messages in the Story
    var totalMessages = 0
        protected set

    //Progress when story is being read
    var progress = 0
        private set

    //Date of publication; Can produce null pointer exception
    var publicationDate: Date? = null

    // Start / Reset Story progress
    fun startStory() {
        currentMessage = storyStart
        progress = 1
    }

    /*
    * Will store next message in currentMessage
    * Returns true if a next message exists
    * false if next message is null
    */
    fun nextMessage(): Boolean {
        if (hasNextMessage()) {
            currentMessage = currentMessage!!.nextMessage
            progress++
            return true
        }
        return false
    }

    //Will check if Story has a next message in the list
    fun hasNextMessage(): Boolean {
        return currentMessage!!.nextMessage != null
    }
}